adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+
+The table below shows a proposal for SYCL supported CUDA extensions. This should
+be populated by other members of the SYCL community.
+
+[[table.extensionsupport]]
+.SYCL support for CUDA 11.3 extensions
+[width="100%",options="header",cols="35%,35%,15%, 15"]
+|====
+| CUDA Extension                            | SYCL Aspect   | Feature Test Macro               | Required Compute Capability 
+|====
+
+[[sub:cuda:builtin-kernel-functions]]
+=== Built-in Kernel Functions
+The CUDA backend specification currently does not define any built-in kernel 
+functions.
+
+[[sub:cuda:error_handling]]
+=== Error Handling
+
+SYCL uses `sycl::errc` as an enum class to hold error codes. These error
+codes may originate in the SYCL runtime or be passed from other runtimes to
+the SYCL runtime. When a `sycl::exception` is thrown, the `sycl::errc` can
+be queried using the exception's `.code()` method. Possible values for 
+`sycl::errc` include: `success`, `runtime`, `memory_allocation`, and more.
+
+If there is a CUDA driver API error associated with an exception triggered, then the
+CUDA error code can be obtained by the free function `CUresult sycl::cuda::get_error_code(sycl::exception&)`. In the case where there is
+no CUDA error associated with the exception triggered, the CUDA error
+code will be `CUDA_SUCCESS`.
+
+Most of the SYCL error codes that form sycl::errc are specifically defined as errors thrown during calls to the SYCL API or SYCL runtime. There are also some cases of sycl::errc which cover errors thrown during the compilation or execution of device code.
+It is suitable to map CUDA errors to such cases, such that an exception, "cuda_exception", that was created due to a CUDA error, may, upon execution of `cuda_exception.code()`, return a `std::error_code` relating to the `sycl::errc` case that the CUDA error maps to; whilst `sycl::cuda::get_error_code(cuda_exception)` will return the original CUDA error code.
+
+The relevant `sycl::errc` cases and the CUDA errors that they may be mapped from are listed below.
+
+==== build
+
+`sycl::errc::build` is defined as:
+
+_Error from an online compile or
+link operation when compiling,
+linking, or building a kernel bundle for a device._
+
+which may be mapped from `CUDA_ERROR_NO_BINARY_FOR_GPU`, `CUDA_ERROR_JIT_COMPILER_NOT_FOUND`, `CUDA_ERROR_INVALID_PTX`, `CUDA_ERROR_UNSUPPORTED_PTX_VERSION`, `CUDA_ERROR_SHARED_OBJECT_INIT_FAILED`, `CUDA_ERROR_SHARED_OBJECT_SYMBOL_NOT_FOUND`.
+
+==== memory_allocation
+
+`sycl::errc::memory_allocation` is defined as:
+
+_Error on memory allocation on the
+SYCL device for a SYCL kernel._
+
+which may be mapped from `CUDA_ERROR_OUT_OF_MEMORY`.
+
+==== kernel_argument
+
+`sycl::errc::kernel_argument` is defined as:
+
+_The application has passed an invalid argument to a SYCL kernel
+function. This includes captured
+variables if the SYCL kernel function is a lambda function._
+
+which may be mapped from `CUDA_ERROR_NOT_FOUND`.
```

Comment:
I wonder if we want this part, though.  Do we really want to list the specific CUDA error codes that could correlate to each SYCL `errc`?  For example, do you want the CTS to fail if `cuda::get_error_code()` returns some CUDA error that is *not* in this list?  I think it might be better to leave this mapping as an implementation detail.

Reply:
There is probably a larger issue here about which error codes should be listed in the SYCL specification.

One option is that the spec should only list error codes that are specified to be thrown someplace in the core spec.  For example, `errc::build` is one of the error codes listed for the kernel bundle APIs, so the core spec needs to list `build` as one of the error codes.  By this rationale, we would remove any error codes that are not specifically called out anywhere in the spec.  (This is the topic of internal issue 512.)

For backend-specific errors, we would throw a SYCL exception with some generic SYCL error code (e.g. `errc::runtime`), and then it would be up to the application to use `sycl::cuda::get_error_code()` in order to get the CUDA specific error.  This way, we do not need to worry about which SYCL error corresponds to each backend error, and we don't need to add new SYCL errors whenever a new backend has an error that doesn't map well.

Reply:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.